pkg/proxy/ipvs/ipset.go

```diff
@@ -102,6 +103,20 @@ func NewIPSet(handle utilipset.Interface, name string, setType utilipset.Type, i
 	hashFamily := utilipset.ProtocolFamilyIPV4
 	if isIPv6 {
 		hashFamily = utilipset.ProtocolFamilyIPV6
+		// In dual-stack both ipv4 and ipv6 ipset's can co-exist. To
+		// ensure unique names the prefix for ipv6 is changed from
+		// "KUBE-" to "KUBE-6-". The "KUBE-" prefix is kept for
+		// backward compatibility. The maximum name lenght of an ipset
+		// is 31 characters which must be taken into account.  The
+		// ipv4 names are not altered to minimize the risk for
+		// problems on upgrades.
```

Comment:
That means a cluster running `IPv6` and is being converted to `Dualstack` will break right? because ipv6 rules will be in `kube-` prefix. Unless we ensure that flush all existing rules if we detect this case.

Reply:
Yes, I fear that is the case. An ipv6-only cluster can be upgraded (even is dual-stack is not enabled) with some rolling-reboot. But it should be possible to migrate to dual-stack so I will look into this.

I can probably feature-gate this, but that would only postpone the problem until dual-stack GA I think, so I would not like to do that.

I have assumed that since ipv6-only stayed in "alpha" there should not be any clusters in production (or at least very few), so this backward incompatibility could be acceptable.

Reply:
The "KUBE-" prefix is kept I hoped clean-up functions would use that, but I have not checked. But if not, it can be updated.

Reply:
What about this: https://unix.stackexchange.com/questions/134578/is-there-a-way-to-match-an-inet-and-inet6-ip-set-in-a-single-rule

It seems like we can create new per-family sets (KUBE4-* and KUBE6-*) which hold the IPs and a union set with the old names KUBE-* which holds the other sets, and naive testing shows that iptables seems to do the right thing...

```
$ ipset create test4 hash:ip

$ ipset add test4 1.2.3.4

$ ipset create test6 hash:ip family inet6

$ ipset add test6 1:2:3:4::1

$ ipset create test list:set

$ ipset add test test4

$ ipset add test test6

$ iptables -t nat -A OUTPUT -m set --match-set test4 dst -j LOG --log-prefix "TEST4   "

$ iptables -t nat -A OUTPUT -m set --match-set test6 dst -j LOG --log-prefix "TEST6   "
iptables v1.6.0: The protocol family of set test6 is IPv6, which is not applicable.

$ iptables -t nat -A OUTPUT -m set --match-set test dst -j LOG --log-prefix "TEST    "

$ curl 1.2.3.4

$ dmesg
[4231193.603254] TEST4   IN= OUT=eth0 SRC=10.128.0.27 DST=1.2.3.4 LEN=60 TOS=0x00 PREC=0x00 TTL=64 ID=17500 DF PROTO=TCP SPT=43710 DPT=80 WINDOW=28400 RES=0x00 SYN URGP=0 
[4231193.618610] TEST   IN= OUT=eth0 SRC=10.128.0.27 DST=1.2.3.4 LEN=60 TOS=0x00 PREC=0x00 TTL=64 ID=17500 DF PROTO=TCP SPT=43710 DPT=80 WINDOW=28400 RES=0x00 SYN URGP=0